bespokesynth: update to latest commit

PowerUser64 opened the pull request:
###### Description of changes

The developer of bespoke synth unfortunately doesn't have much time to devote to creating version number releases for the project, so the current "stable" release has gone untouched for quite some time. Currently, the latest commit/nightly release is far more stable and featureful than the current stable release. This PR updates the BespokeSynth package to the latest commit.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

PowerUser64 committed:
bespokesynth: 1.1.0 -> unstable-2023-08-17

Co-authored-by: Christoph Neidahl <christoph.neidahl@gmail.com>
Co-authored-by: Tobias Bora <tobias.bora.list@gmail.com>

tobiasBora commented:
Nice, thanks for helping us with this! Do you think you could also take this opportunity to fix the way we patch the program, by directly using `rpath` instead of `LD_LIBRARY_PATH` as we were discussing [here](https://discourse.nixos.org/t/making-a-development-flake-for-bespoke-synth/29988/8), in order to provide a smoother experience for devs? Should be just a matter of removing:
```
--prefix LD_LIBRARY_PATH : '${lib.makeLibraryPath [
          libXrandr
          libXinerama
          libXcursor
          libXScrnSaver
        ]}
```
and adding:
```
NIX_LDFLAGS = "-rpath ${lib.makeLibraryPath (with xorg; [ libX11 libXrandr libXinerama libXext libXcursor libXScrnSaver ])} $NIX_LDFLAGS";
dontPatchELF = true; # needed or nix will try to optimize the binary by removing "useless" rpath
```

Thanks!

OPNA2608 reviewed (commented):
Breaks Darwin builds as-is.

![Screenshot_macOS_2023-07-10_22:53:54](https://github.com/NixOS/nixpkgs/assets/23431373/1e1bad38-ebbb-42bd-a92c-3ba842e4fa17)

PowerUser64 commented:
It seems that trying to pass through `NIX_LDFLAGS` in the same way we were doing in the flake causes it to fail compiling a basic test program because the variable isn't expanded. I've gone removed `$NIX_LDFLAGS` from the end of the `NIX_LDFLAGS = ...` line.

OPNA2608 reviewed (changes requested):
For fixing Darwin support, please add `MetalKit` to the Darwin-specific `buildInputs` and, in `pkgs/top-level/all-packages.nix`, change
```nix
  bespokesynth = callPackage ../applications/audio/bespokesynth {
    inherit (darwin.apple_sdk.frameworks) Accelerate Cocoa WebKit CoreServices CoreAudioKit IOBluetooth;
  };
```
to
```nix
  bespokesynth = darwin.apple_sdk_11_0.callPackage ../applications/audio/bespokesynth {
    inherit (darwin.apple_sdk_11_0.frameworks) Accelerate Cocoa WebKit CoreServices CoreAudioKit IOBluetooth MetalKit;
  };
```

PowerUser64 commented:
@OPNA2608 does this do it?

tobiasBora commented:
I can confirm that it works for what I tested on NixOs! (VST version etc), good for me if it also works on MacOs!

PowerUser64 commented:
Quick question, I'm going to try to keep this up-to-date as much as possible. I don't want you all to hate me because I'm creating/updating PR's every time there's a new bespoke nightly. Should I try to limit PR's to one per week or something? How should we handle this?

For reference, bespoke nightly comes out about one per commit usually. This can as often as anywhere from one every two weeks or (especially recently), one or more per day. I always run the latest version with `overrideAttrs` in my system config, so I should always know whether it's broken on Linux at least.

tobiasBora commented on pkgs/applications/audio/bespokesynth/default.nix:
```diff
@@ -133,23 +131,20 @@ stdenv.mkDerivation rec {
         ]}'
     '';
 
+  NIX_LDFLAGS = "-rpath ${lib.makeLibraryPath ([ libX11 libXrandr libXinerama libXext libXcursor libXScrnSaver ])}";
```
Why should it be `env.`? Any variable in a derivation is automatically propagated to the environment variables… and in nixpkgs all but 2 programs use `env.NIX_LDFLAGS`:
```
$ rg "NIX_LDFLAGS =" | wc -l
193
$ rg "env.NIX_LDFLAGS =" | wc -l 
2
```

OPNA2608 replied:
I know `NIX_CFLAGS_COMPILE` was recently migrated treewide to `env.NIX_CFLAGS_COMPILE` for complex reasons: https://github.com/NixOS/nixpkgs/pull/217206

Not sure if the same was done for `NIX_LDFLAGS`, if `env.NIX_LDFLAGS` works then I'd prefer that.

tobiasBora replied:
Oh good to know, sorry I was not aware of this new policy.

tobiasBora commented:
@PowerUser64 Well, I am not sure if nixpkgs is well suited for nightly versions (but great to see that this project is getting much attention). Since the review process can be quite long and energy consuming (we need to test across several systems like MacOs to be sure there is no breakage, find a nix maintainer to merge it, wait for hydra to build it etc…) I would say that adding one release per day, or even one per month for a basic software is too much. Every 3 months is maybe more reasonable, unless a really important bug/security issue is fixed (but you might ask for confirmation to more experienced users, e.g. in matrix or discourse).

If you want to deal with nightly build, I would rather advice you create your own repository (It is what people are doing for instance [emacs](https://github.com/nix-community/emacs-overlay)) and to integrate it into [NUR](https://github.com/nix-community/NUR), or setup your own repository.

tobiasBora commented:
Great thanks, looks good to me!

PowerUser64 commented:
> Every 3 months is maybe more reasonable, unless a really important bug/security issue is fixed.
>
> If you want to deal with nightly build, I would rather advice you create your own repository and to integrate it into [NUR](https://github.com/nix-community/NUR)

Good to know. I'll try and limit updates to about one per 3 months, or when an official version number release comes out. There's supposedly going to be a major version update in the next few weeks, so I will probably do one then, and then who knows when the next official update will be.

wackbyte commented:
[v1.2.0](https://github.com/BespokeSynth/BespokeSynth/releases/tag/v1.2.0) has just released :)

PowerUser64 commented:
Woohoo! New commit coming in hot!

OPNA2608 commented on pkgs/applications/audio/bespokesynth/default.nix:
```diff
@@ -133,23 +131,27 @@ stdenv.mkDerivation rec {
         ]}'
     '';
 
+  env.NIX_LDFLAGS = "-rpath ${lib.makeLibraryPath ([
```
```suggestion
  env.NIX_LDFLAGS = lib.optionalString stdenv.hostPlatform.isLinux "-rpath ${lib.makeLibraryPath ([
```

Otherwise this pulls in all of the X11 stuff on Darwin.
![image](https://github.com/NixOS/nixpkgs/assets/23431373/bba8cdbe-4eff-4f16-9bdd-bb48ae4450ab)

OPNA2608 reviewed (approved):
Result of `nixpkgs-review pr 242698` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>bespokesynth</li>
    <li>bespokesynth-with-vst2</li>
  </ul>
</details>

---

Result of `nixpkgs-review pr 242698` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>bespokesynth</li>
    <li>bespokesynth-with-vst2</li>
  </ul>
</details>

---

Works on both systems, LGTM.

luzpaz commented:
Any progress here ?

OPNA2608 commented on pkgs/applications/audio/bespokesynth/default.nix:
```diff
@@ -58,28 +59,24 @@ let
     repo = pname;
-    rev = "v${version}";
-    sha256 = "sha256-PN0Q6/gI1PeMaF/8EZFGJdLR8JVHQZfWunAhOIQxkHw=";
+    rev = "71134ce432dd15ae773cd49ac7c6e436d37e8a46";
+    hash = "sha256-pEifqbzJYI0EnGRQL0a0hzNkUw4+ZE6mrMGUVngzTZY=";
```
I didn't pick up on the fact that when you force-pushed from bda87f7beedf33ce00bad49ad34025ccea373c02 to 179c1e2b601448d2a14f72218b8b17a0046cc5b8 , you went back to using an unstable commit without marking the `version` as such. 

https://github.com/BespokeSynth/BespokeSynth/commit/71134ce432dd15ae773cd49ac7c6e436d37e8a46 != https://github.com/BespokeSynth/BespokeSynth/releases/tag/v1.2.0 . 

If this is still supposed to be `version = "1.2.0"` then please revert the `rev` line to the format that was used before and change `hash` to what the actual hash of that version is:

```suggestion
    rev = "v${version}";
    hash = "sha256-baBZBRK8xbtnj/MJaTA036exLjvvucfrsk/Ue5oiG7c=";
```

PowerUser64 replied:
Sorry, I should have left a comment saying I did so. I updated to the latest commit because I learned the stable version had broken jedi functionality (python autocomplete) on MacOS. Since I don't know how long it was going to be until the next release (or opportunity to update this package), I went ahead and updated to the latest commit (which fixed this). I skipped updating the version number because I saw it as basically being a hotfix to v1.2.0, but I should probably mark it as unstable. Should I mark it as `v1.2.0-unstable` or would the date of the commit be better?

tobiasBora replied:
It seems like most of the packages opted for `v1.2.0+unstable=2023-07-20`:
```
pkgs/tools/misc/tty-clock/default.nix
5:  version = "2.3+unstable=2021-04-07";

pkgs/tools/misc/dialogbox/default.nix
11:  version = "1.0+unstable=2020-11-16";

pkgs/applications/emulators/bsnes/higan/default.nix
25:  version = "115+unstable=2021-08-18";

pkgs/applications/emulators/emu2/default.nix
8:  version = "0.pre+unstable=2021-09-22";

pkgs/applications/graphics/gimp/plugins/default.nix
225:    version = "2.2+unstable=2021-12-03";

pkgs/applications/graphics/inkscape/extensions/applytransforms/default.nix
9:  version = "0.pre+unstable=2021-05-11";

pkgs/applications/terminal-emulators/st/siduck76-st.nix
15:  version = "0.pre+unstable=2021-08-20";

pkgs/applications/terminal-emulators/st/lukesmithxyz-st/default.nix
15:  version = "0.pre+unstable=2021-08-10";

…
```

OPNA2608 replied:
> I learned the stable version had broken jedi functionality (python autocomplete) on MacOS

If it's easy to figure out which commits fixed this, then you could pull them in and apply them onto 1.2.0:

```nix
patches = [
  # Fixes jedi functionality on Darwin
  # Remove when version > 1.2.0
  (fetchpatch {
    url = "https://github.com/BespokeSynth/bespokesynth/commit/whatevercommitfixesthis.patch";
    hash = "sha256-whateverhashisneededhere=";
  })
];
```

> I skipped updating the version number because I saw it as basically being a hotfix to v1.2.0, but I should probably mark it as unstable

Unless it's an actual release with a tag (or in some other form actually treated by upstream as a release, with or without tag), it should be treated as an unstable commit instead of a release. For pulling in a fix into a release, see above.

> Should I mark it as `v1.2.0-unstable` or would the date of the commit be better?

The above would be preferable IMO. If you want to ditch the applying-patches way, then back to what [the manual](https://nixos.org/manual/nixpkgs/stable/#sec-package-naming) says unstable versions should use: `unstable-<YYYY-MM-DD of commit>`

OPNA2608 replied:
> It seems like most of the packages opted for `v1.2.0+unstable=2023-07-20`

This versioning pattern is from [a not-yet-accepted RFC](https://github.com/NixOS/rfcs/pull/107), dates used in them is ~ when the RFC was first opened. And they don't align with what the RFC currently suggests to use: `3.0:date=2008-12-04`.

I'd prefer to wait until *any* consensus is reached on that RFC & it gets accepted before using its pattern, as right now it conflicts with what the manual has to say on this.

PowerUser64 replied:
Okay! When I get the chance, I'll update the version to `unstable-2023-08-17` and the commit hash to [`c6b141`](https://github.com/BespokeSynth/BespokeSynth/commit/c6b1410afefc8b0b9aeb4aa11ad5c32651879c9f) (slightly behind `main`). I think that's a good stable stopping place because most things added after it are features.
Does this sound good?

PowerUser64 replied:
I'll also ask awwbees about making that into `v1.2.1`, as I think that would be a good place for a new version.

OPNA2608 reviewed (changes requested):
Ah sorry, seems I didn't review this too well…

In addition to the suggestion, can you make the commit title follow the [commit conventions](https://github.com/NixOS/nixpkgs/blob/63ad63874330a6cb946837a545e207183411891a/CONTRIBUTING.md#commit-conventions)? It should be:

`bespokesynth: 1.1.0 -> 1.2.0`

…if you still only want to bump to the next stable version, instead of an unstable commit.

Once it LGTM this time, I'll try to find someone with merge permissions to get this in.

PowerUser64 commented:
Okay, I've changed the version to `unstable-2023-08-17` and updated to the commit I said I was going to update to (https://github.com/BespokeSynth/BespokeSynth/commit/c6b1410afefc8b0b9aeb4aa11ad5c32651879c9f).

PowerUser64 commented:
Realized I forgot to update the commit message - just did it.

`bespokesynth: 1.1.0 -> unstable-2023-08-17`

PowerUser64 commented:
What can I do to help get this merged?

tobiasBora commented:
You can drop a message on https://discourse.nixos.org/t/prs-already-reviewed/2617/726

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1088
